src/parser/ExpressionParser.mjs

```diff
@@ -250,146 +250,217 @@ export class ExpressionParser extends FunctionParser {
   //   CoalesceExpression
   //   BitwiseORExpression
   parseShortCircuitExpression() {
-    // Start parse at BIT_OR, right above AND/OR/NULLISH
-    const expression = this.parseBinaryExpression(TokenPrecedence[Token.BIT_OR]);
-    switch (this.peek().type) {
-      case Token.AND:
-      case Token.OR:
-        // Drop into normal binary chain starting at OR
-        return this.parseBinaryExpression(TokenPrecedence[Token.OR], expression);
-      case Token.NULLISH: {
-        let x = expression;
-        while (this.eat(Token.NULLISH)) {
-          const node = this.startNode();
-          node.CoalesceExpressionHead = x;
-          node.BitwiseORExpression = this.parseBinaryExpression(TokenPrecedence[Token.BIT_OR]);
-          x = this.finishNode(node, 'CoalesceExpression');
-        }
-        return x;
+    let result = this.parseLogicalORExpression();
+    while (this.test(Token.NULLISH)) {
+      const { type } = result;
+      if (type === 'LogicalANDExpression' || type === 'LogicalORExpression') {
+        this.raise(
+          'MixingLogicalOperators', this.peek(),
+          result.operator, this.peek().value,
+        );
       }
-      default:
-        return expression;
+      const node = this.startNode(result);
+      node.CoalesceExpressionHead = result;
+      node.operator = this.next().value;
+      node.BitwiseORExpression = this.parseBitwiseORExpression();
+      result = this.finishNode(node, 'CoalesceExpression');
     }
+    return result;
   }
 
-  parseBinaryExpression(precedence, x) {
-    if (!x) {
-      if (this.test(Token.PRIVATE_IDENTIFIER)) {
-        x = this.parsePrivateIdentifier();
-        if (!this.test(Token.IN)) {
-          this.raise('UnexpectedToken');
-        }
-        this.scope.checkUndefinedPrivate(x);
-      } else {
-        x = this.parseUnaryExpression();
-      }
+  // parse a sequence of same-precedence left-associative binary operators
+  parseBinaryChain(parseArg, nodeType, leftKey, rightKey, ...operators) {
+    let result = this[parseArg]();
+    while (operators.includes(this.peek().type)) {
```

Comment:
the rest arg and array.includes are not ideal... i wonder if we could create a closure which has a `Set` of valid operators instead.

Reply:
If you mean not ideal for speed, the array has at most 6 elements for `RelationalExpression`, and 1-4 elements for all others. I'm pretty sure the performance difference between `array.includes` and `set.has` is not worth having these sets defined somewhere else.

```
$ pnpx benchmark -e 'function inSpread(x, ...ys) { return ys.includes(x); } for (let i=0; i<1e6; ++i) inSpread(7, 1);'
✔ eval-1 x 130 ops/sec ±0.61% (79 runs sampled)
$ pnpx benchmark -e 'function inSet(x, ys) { return ys.has(x); } for (let i=0, set=new Set([1]); i<1e6; ++i) inSet(7, set);'
✔ eval-1 x 94.11 ops/sec ±0.56% (77 runs sampled)

$ pnpx benchmark -e 'function inSpread(x, ...ys) { return ys.includes(x); } for (let i=0; i<1e6; ++i) inSpread(7, 1,2,3);'
✔ eval-1 x 93.99 ops/sec ±0.67% (77 runs sampled)
$ pnpx benchmark -e 'function inSet(x, ys) { return ys.has(x); } for (let i=0, set=new Set([1,2,3]); i<1e6; ++i) inSet(7, set);'
✔ eval-1 x 82.82 ops/sec ±0.47% (69 runs sampled)

$ pnpx benchmark -e 'function inSpread(x, ...ys) { return ys.includes(x); } for (let i=0; i<1e6; ++i) inSpread(7, 1,2,3,4,5,6);'
✔ eval-1 x 65.66 ops/sec ±0.65% (66 runs sampled)
$ pnpx benchmark -e 'function inSet(x, ys) { return ys.has(x); } for (let i=0, set=new Set([1,2,3,4,5,6]); i<1e6; ++i) inSet(7, set);'
✔ eval-1 x 94.60 ops/sec ±0.34% (77 runs sampled)
```